go/vt/vtgr/controller/diagnose_test.go

```diff
@@ -0,0 +1,615 @@
+					DoAndReturn(func(alias string, target *inst.InstanceKey) (*db2.GroupView, error) {
+						if target.Hostname == "" || target.Port == 0 {
+							return nil, errors.New("invalid mysql instance key")
+						}
+						return db2.BuildGroupView(alias, "group", testHost, testPort0, false, []db2.TestGroupState{
+							{MemberHost: testHost, MemberPort: strconv.Itoa(testPort0), MemberState: "ONLINE", MemberRole: "PRIMARY"},
+							{MemberHost: testHost, MemberPort: strconv.Itoa(testPort1), MemberState: "ONLINE", MemberRole: "SECONDARY"},
+							{MemberHost: testHost, MemberPort: strconv.Itoa(testPort2), MemberState: "ONLINE", MemberRole: "SECONDARY"},
+						}), nil
+					}).
+					AnyTimes()
+				tmc.
+					EXPECT().
+					Ping(gomock.Any(), &topodatapb.Tablet{
+						Alias:               tablet.Alias,
+						Hostname:            tablet.Hostname,
+						Keyspace:            tablet.Keyspace,
+						Shard:               tablet.Shard,
+						Type:                tablet.Type,
+						Tags:                tablet.Tags,
+						MysqlHostname:       tablet.MysqlHostname,
+						MysqlPort:           tablet.MysqlPort,
+						MasterTermStartTime: tablet.MasterTermStartTime,
+					}).
+					DoAndReturn(func(_ context.Context, t *topodatapb.Tablet) error {
+						if !response.pingable {
+							return errors.New("unreachable")
+						}
+						return nil
+					}).
+					AnyTimes()
+			}
+			ts.
+				EXPECT().
+				GetTabletMapForShardByCell(gomock.Any(), gomock.Eq("ks"), gomock.Eq("0"), gomock.Any()).
+				Return(tablets, nil)
+
+			ctx := context.Background()
+			cfg := &config.VTGRConfig{GroupSize: diagnoseGroupSize, MinNumReplica: 2, BackoffErrorWaitTimeSeconds: 1, BootstrapWaitTimeSeconds: 1}
+			shard := NewGRShard("ks", "0", nil, tmc, ts, db, cfg, testPort0)
+			shard.refreshTabletsInShardLocked(ctx)
+			diagnose, err := shard.Diagnose(ctx)
+			assert.Equal(t, expected, diagnose)
+			if tt.errMessage == "" {
+				assert.NoError(t, err)
+			} else {
+				assert.Error(t, err)
+				assert.True(t, strings.Contains(err.Error(), tt.errMessage), err.Error())
+			}
+		})
+	}
+}
+
+func TestMysqlIssueDiagnoses(t *testing.T) {
+	cfg := &config.VTGRConfig{GroupSize: diagnoseGroupSize, MinNumReplica: 2, BackoffErrorWaitTimeSeconds: 1, BootstrapWaitTimeSeconds: 1}
+	disableProtectionCfg := &config.VTGRConfig{GroupSize: diagnoseGroupSize, MinNumReplica: 2, DisableReadOnlyProtection: true, BackoffErrorWaitTimeSeconds: 1, BootstrapWaitTimeSeconds: 1}
+	type data struct {
+		alias      string
+		groupName  string
+		readOnly   bool
+		groupInput []db2.TestGroupState
+		ttype      topodatapb.TabletType
+	}
+	var sqltests = []struct {
+		name          string
+		expected      DiagnoseType
+		errMessage    string
+		config        *config.VTGRConfig
+		inputs        []data
+		removeTablets []string // to simulate missing tablet in topology
+	}{
+		{name: "healthy shard", expected: DiagnoseTypeHealthy, errMessage: "", inputs: []data{
```

Comment:
I didn't verify this, but I'm assuming all DiagnoseTypes are covered by this unit test.

Reply:
yeah, we have all 10 types (expect unknown) covered:
```
$ grep "expected: DiagnoseType" go/vt/vtgr/controller/diagnose_test.go | awk -F','   '{print $2}' | sort | uniq | wc -l
      10
```